src/cpp/tensorpack.cpp

```diff
@@ -30,152 +30,153 @@
+// the tensor pack. This function will cast the TensorBase to the correct
+// Tensor<T> type.
 void TensorPack::add_tensor(TensorBase* tensor)
 {
     std::string name = tensor->name();
 
-    if(name.size()==0)
-        throw std::runtime_error("The tensor name must "\
-                                 "be greater than 0.");
+    if (name.size() == 0)
+        throw std::runtime_error("The tensor name must be nonempty.");
 
-    this->_tensorbase_inventory[name] = tensor;
-    this->_all_tensors.push_front(tensor);
-    return;
+    _tensorbase_inventory[name] = tensor;
+    _all_tensors.push_front(tensor);
 }
 
+// Return a TensorBase pointer based on name.
 TensorBase* TensorPack::get_tensor(const std::string& name)
 {
-    TensorBase* ptr = this->_tensorbase_inventory.at(name);
-    return ptr;
+    return _tensorbase_inventory.at(name);
 }
 
+// Retrieve a pointer to the tensor data memory space
 void* TensorPack::get_tensor_data(const std::string& name)
 {
-    TensorBase* ptr = this->_tensorbase_inventory.at(name);
-    return ptr->data();
+    TensorBase* ptr = _tensorbase_inventory.at(name);
+    return (ptr == NULL ? ptr : ptr->data());
 }
 
+// Check whether a tensor with a given name exists in the TensorPack
 bool TensorPack::tensor_exists(const std::string& name)
 {
-    return (this->_tensorbase_inventory.count(name)>0);
+    return (_tensorbase_inventory.count(name) > 0);
 }
 
+// Retrieve an iterator pointing to the first Tensor
 TensorPack::tensorbase_iterator TensorPack::tensor_begin()
 {
-    return this->_all_tensors.begin();
+    return _all_tensors.begin();
 }
 
+// Retrieve an iterator pointing to the last Tensor
 TensorPack::tensorbase_iterator TensorPack::tensor_end()
 {
-    return this->_all_tensors.end();
+    return _all_tensors.end();
 }
 
+// Retrieve a const iterator pointing to the first Tensor
 TensorPack::const_tensorbase_iterator TensorPack::tensor_cbegin() const
 {
-    return this->_all_tensors.cbegin();
+    return _all_tensors.cbegin();
 }
 
+// Retrieve a const iterator pointing to the last Tensor
 TensorPack::const_tensorbase_iterator TensorPack::tensor_cend() const
 {
-    return this->_all_tensors.cend();
+    return _all_tensors.cend();
 }
 
+// Copy the tensor inventory from one TensorPack to this TensorPack
 void TensorPack::_copy_tensor_inventory(const TensorPack& tp)
 {
-    typename TensorPack::const_tensorbase_iterator it =
-        tp.tensor_cbegin();
-    typename TensorPack::const_tensorbase_iterator it_end =
-        tp.tensor_cend();
-
-    TensorBase* ptr;
-    while(it!=it_end) {
-        ptr = (*it)->clone();
-        this->_all_tensors.push_front(ptr);
-        this->_tensorbase_inventory[ptr->name()] = ptr;
-        it++;
+    // Check for self-copy
+    if (this == &tp)
+        return;
+
+    typename TensorPack::const_tensorbase_iterator it = tp.tensor_cbegin();
+    for ( ; it != tp.tensor_cend(); it++) {
+        TensorBase* ptr = (*it)->clone();
+        if (ptr == NULL)
```

Comment:
We should check if clone() can return ``NULL``, and if it can, I think we should throw an error at the end of clone``clone()``.  Also, if ``(*it)`` is a valid tensor and we skip it because clone is null, the copy operation failed and we should throw an error, not skip it and have the tensorpacks be difference.